UnsupportedOperationException while running a QuarkusUnitTest with DevServices

### Describe the bug

Having a test like the following: 

```java
package io.quarkus.registry.app.services;

import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;

import io.quarkus.test.QuarkusUnitTest;
import org.jboss.shrinkwrap.api.ShrinkWrap;
import org.jboss.shrinkwrap.api.asset.StringAsset;
import org.jboss.shrinkwrap.api.spec.JavaArchive;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;

import static io.restassured.RestAssured.given;
import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.CoreMatchers.is;

public class RegistryDescriptorMavenResourceTest {

    @RegisterExtension
    static final QuarkusUnitTest config = new QuarkusUnitTest()
            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                    .addAsResource(new StringAsset("quarkus.registry.groupId=foo"), "application.properties"));

    @Test
    void should_use_custom_group_id() {
        given()
                .get("/maven/foo/quarkus-registry-descriptor/1.0-SNAPSHOT/quarkus-registry-descriptor-1.0-SNAPSHOT.json")
                .then()
                .statusCode(200)
                .header(HttpHeaders.CONTENT_TYPE, containsString(MediaType.APPLICATION_JSON))
                .body("descriptor.artifact", is("foo:quarkus-registry-descriptor::json:1.0-SNAPSHOT"),
                        "platforms.artifact", is("foo:quarkus-platforms::json:1.0-SNAPSHOT"));

    }
}
```

The following error is thrown when trying to run the test:

```java
Caused by: java.lang.UnsupportedOperationException
	at io.quarkus.test.junit.RunningAppConfigResolver$1.unwrap(RunningAppConfigResolver.java:57)
	at io.quarkus.runtime.configuration.ConfigUtils.isPropertyPresent(ConfigUtils.java:208)
	at io.quarkus.datasource.deployment.spi.DevServicesDatasourceConfigurationHandlerBuildItem$2.test(DevServicesDatasourceConfigurationHandlerBuildItem.java:68)
	at io.quarkus.datasource.deployment.spi.DevServicesDatasourceConfigurationHandlerBuildItem$2.test(DevServicesDatasourceConfigurationHandlerBuildItem.java:64)
	at io.quarkus.datasource.deployment.devservices.DevServicesDatasourceProcessor.startDevDb(DevServicesDatasourceProcessor.java:234)
	at io.quarkus.datasource.deployment.devservices.DevServicesDatasourceProcessor.launchDatabases(DevServicesDatasourceProcessor.java:114)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2442)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1476)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

```

### Expected behavior

The test would run

### Actual behavior

Test fails with `UnsupportedOperationException`

### How to Reproduce?

Run the test above with a project containing:

```xml
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-hibernate-orm-panache</artifactId>
    </dependency>
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-jdbc-postgresql</artifactId>
    </dependency>
```

### Output of `uname -a` or `ver`

Fedora 34

### Output of `java -version`

11

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.1.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.1

### Additional information

_No response_

--- a/test-framework/junit5-internal/src/main/java/io/quarkus/test/ExclusivityChecker.java
+++ b/test-framework/junit5-internal/src/main/java/io/quarkus/test/ExclusivityChecker.java
@@ -1,0 +1,23 @@
+package io.quarkus.test;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+public class ExclusivityChecker {
+    public static final String IO_QUARKUS_TESTING_TYPE = "io.quarkus.testing.type";
+
+    public static void checkTestType(ExtensionContext extensionContext, Class<?> current) {
+        ExtensionContext.Store store = extensionContext.getRoot().getStore(ExtensionContext.Namespace.GLOBAL);
+
+        Class<?> testType = store.get(IO_QUARKUS_TESTING_TYPE, Class.class);
+        if (testType != null) {
+            if (testType != QuarkusUnitTest.class && testType != QuarkusDevModeTest.class
+                    && testType != QuarkusProdModeTest.class) {
+                throw new IllegalStateException(
+                        "Cannot mix both " + current.getName() + " based tests and " + testType.getName()
+                                + " based tests in the same run");
+            }
+        } else {
+            store.put(IO_QUARKUS_TESTING_TYPE, current);
+        }
+    }
+}

--- a/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java
+++ b/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java
@@ -188,6 +188,7 @@ public class QuarkusDevModeTest
         if (archiveProducer == null) {
             throw new RuntimeException("QuarkusDevModeTest does not have archive producer set");
         }
+        ExclusivityChecker.checkTestType(extensionContext, QuarkusDevModeTest.class);
 
         if (logFileName != null) {
             PropertyTestUtil.setLogFileProperty(logFileName);

--- a/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusProdModeTest.java
+++ b/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusProdModeTest.java
@@ -334,6 +334,7 @@ public class QuarkusProdModeTest
     @Override
     public void beforeAll(ExtensionContext extensionContext) throws Exception {
         ensureNoInjectAnnotationIsUsed(extensionContext.getRequiredTestClass());
+        ExclusivityChecker.checkTestType(extensionContext, QuarkusProdModeTest.class);
 
         originalHandlers = rootLogger.getHandlers();
         rootLogger.addHandler(inMemoryLogHandler);

--- a/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java
+++ b/test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusUnitTest.java
@@ -401,6 +401,9 @@ public class QuarkusUnitTest
             PropertyTestUtil.setLogFileProperty();
         }
         ExtensionContext.Store store = extensionContext.getRoot().getStore(ExtensionContext.Namespace.GLOBAL);
+
+        ExclusivityChecker.checkTestType(extensionContext, QuarkusUnitTest.class);
+
         TestResourceManager testResourceManager = (TestResourceManager) store.get(TestResourceManager.class.getName());
         if (testResourceManager == null) {
             testResourceManager = new TestResourceManager(extensionContext.getRequiredTestClass());

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -124,6 +124,7 @@ public class QuarkusTestExtension
     protected static final String TEST_LOCATION = "test-location";
     protected static final String TEST_CLASS = "test-class";
     public static final String QUARKUS_TEST_HANG_DETECTION_TIMEOUT = "quarkus.test.hang-detection-timeout";
+    public static final String IO_QUARKUS_TESTING_TYPE = "io.quarkus.testing.type";
 
     private static boolean failedBoot;
 
@@ -659,6 +660,15 @@ public class QuarkusTestExtension
     private ExtensionState ensureStarted(ExtensionContext extensionContext) {
         ExtensionContext root = extensionContext.getRoot();
         ExtensionContext.Store store = root.getStore(ExtensionContext.Namespace.GLOBAL);
+        Class<?> testType = store.get(IO_QUARKUS_TESTING_TYPE, Class.class);
+        if (testType != null) {
+            if (testType != QuarkusTest.class) {
+                throw new IllegalStateException(
+                        "Cannot mix both @QuarkusTest based tests and " + testType.getName() + " based tests in the same run");
+            }
+        } else {
+            store.put(IO_QUARKUS_TESTING_TYPE, QuarkusTest.class);
+        }
         ExtensionState state = store.get(ExtensionState.class.getName(), ExtensionState.class);
         Class<? extends QuarkusTestProfile> selectedProfile = getQuarkusTestProfile(extensionContext);
         boolean wrongProfile = !Objects.equals(selectedProfile, quarkusTestProfile);

